File: 298. Binary Tree Longest Consecutive Sequence/AC_Answer_1.py
# Use DFS and stack

class Solution(object):
    def longestConsecutive(self, root):
        res = 0
        if not root:
            return res
        stack = [(root, 1)]
        while stack:
            node, cnt = stack.pop()
            if node.left:
                stack.append((node.left, cnt+1 if node.left.val == node.val + 1 else 1))
            if node.right:
                stack.append((node.right, cnt+1 if node.right.val == node.val + 1 else 1))
            res = max(res, cnt)
        return res

File: 314. Binary Tree Vertical Order Traversal/AC_Answer_1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        // can not use dfs, use bfs instead
        Map<Integer, List<Integer>> map = new TreeMap<>();
        Queue<TreeNode> stack = new LinkedList<>();
        Queue<Integer> levels = new LinkedList<>();
        stack.add(root);
        levels.add(0);
        while(!stack.isEmpty()){
            TreeNode cur = stack.poll();
            int cur_level = levels.poll();
            if(cur!=null){
                if(!map.containsKey(cur_level)) map.put(cur_level, new ArrayList<>());
                map.get(cur_level).add(cur.val);
                
                if(cur.left!=null){
                    stack.offer(cur.left);
                    levels.offer(cur_level-1);
                }
                
                if(cur.right!=null){
                    stack.offer(cur.right);
                    levels.offer(cur_level+1);
                }
                
            }
            
        }
        return new ArrayList<List<Integer>>(map.values());
    }
}
